drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    /* TODO: protect by spinlock ... */
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+    SocketGet(s);
+
+    if (Direction == DIR_SEND)
+    {
+        /* This will create a tdiIrp: */
+        status =
+            TdiSend(&tdiIrp, s->ConnectionFile, 0, ((char *)BufferData) + Buffer->Offset, Buffer->Length,
+                    NetioComplete, nc);
+    }
+    else
+    {
+        status =
+            TdiReceive(&tdiIrp, s->ConnectionFile, 0, ((char *)BufferData) + Buffer->Offset, Buffer->Length,
+                       NetioComplete, nc);
+    }
+
+    /* TODO: check this again..I would assume if the TDI helper function
+       returns any error, the completion function is not called.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        ExFreePoolWithTag(nc, 'NEIO');
+    }
+
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskSend(_In_ PWSK_SOCKET Socket, _In_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    return WskStreamIo(Socket, Buffer, Flags, Irp, DIR_SEND);
+}
+
+static NTSTATUS WSKAPI
+WskReceive(_In_ PWSK_SOCKET Socket, _In_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    return WskStreamIo(Socket, Buffer, Flags, Irp, DIR_RECEIVE);
+}
+
+static NTSTATUS WSKAPI
+WskDisconnect(_In_ PWSK_SOCKET Socket, _In_opt_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    DbgPrint("WskDisconnect Not implemented\n");
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+
+static struct _WSK_PROVIDER_CONNECTION_DISPATCH TcpDispatch = {
+    .WskControlSocket = WskControlSocket,
+    .WskCloseSocket = WskCloseSocket,
+    .WskBind = WskBind,
+    .WskConnect = WskConnect,
+    .WskGetLocalAddress = WskGetLocalAddress,
+    .WskGetRemoteAddress = WskGetRemoteAddress,
+    .WskSend = WskSend,
+    .WskReceive = WskReceive,
+    .WskDisconnect = WskDisconnect,
+    .WskRelease = WskReleaseTcp,
+};
+
+static NTSTATUS WSKAPI
+WskSocket(PWSK_CLIENT Client,
+          ADDRESS_FAMILY AddressFamily,
+          USHORT SocketType,
+          ULONG Protocol,
+          ULONG Flags,
+          PVOID SocketContext,
+          const VOID * Dispatch,
+          PEPROCESS OwningProcess, PETHREAD OwningThread, PSECURITY_DESCRIPTOR SecurityDescriptor, PIRP Irp)
```

Comment:
```suggestion
WskSocket(
    PWSK_CLIENT Client,
    ADDRESS_FAMILY AddressFamily,
    USHORT SocketType,
    ULONG Protocol,
    ULONG Flags,
    PVOID SocketContext,
    const VOID * Dispatch,
    PEPROCESS OwningProcess, PETHREAD OwningThread, PSECURITY_DESCRIPTOR SecurityDescriptor, PIRP Irp)
```
don't rely on the func names length on indent. notice that the last line gets shorter also.